src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecParametersProvider.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+
+using System.Reflection;
+using System.Text.Json;
+using Nethermind.Config;
+
+namespace Nethermind.Specs.ChainSpecStyle;
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core;
+
+public class ChainSpecParametersProvider : IChainSpecParametersProvider
+{
+    // TODO: test that all IChainSpecEngineParameters have this suffix
+    private const string EngineParamsSuffix = "ChainSpecEngineParameters";
+
+    private readonly Dictionary<string, JsonElement> _chainSpecParameters =
+        new(StringComparer.InvariantCultureIgnoreCase);
+
+    private readonly ConcurrentDictionary<Type, IChainSpecEngineParameters> _instances = new();
+    public string SealEngineType { get; }
+
+    public ChainSpecParametersProvider(Dictionary<string, JsonElement> engineParameters)
+    {
+        foreach (var parameters in engineParameters)
+        {
+            _chainSpecParameters[parameters.Key] = parameters.Value;
+        }
```

Comment:
For shallow copies you can do:

```suggestion
        _chainSpecParameters = new Dictionary<string, JsonElement>(engineParameters, StringComparer.InvariantCultureIgnoreCase);
```